realtek: add support for Linksys LGS310C

plappermaul opened the pull request:
```
Hardware specification
----------------------

* RTL8380M SoC, 1 MIPS 4KEc core @ 500MHz
* 256MB DRAM
* 32MB NOR Flash
* 8 x 10/100/1000BASE-T ports
* 2 x SFP ports
* Power LED, Fault LED
* Reset button on front panel
* UART (115200 8N1) via populated standard pin header marked JP1

TODO: If SFP modules are inserted into Slot 9 and 10 they are detected. No tests have been run yet.

UART pinout
-----------

[o]ooo|JP1
 | ||`------ GND
 | |`------- RX
 | `-------- TX
 `---------- Vcc (3V3)

Installation using OEM webinterface
-----------------------------------

1. Make sure you are running OEM firmware in secondary slot
2. Install squashfs-factory.imag to primary slot by upload via http

Installation using serial interface
-----------------------------------

1. Press "a" "c" "p" during message "Enter correct key to stop autoboot"
2. Load image with "upgrade runtime <TFTP IP>:squashfs-sysupgrade.bin" command
3. Switch to primary slot with "setsys bootpartition 0"
4. Store config with "savesys"
5. Boot the image with `boota` command

Dual-boot with stock firmware using writable u-boot-env 
-------------------------------------------------------

From stock to OpenWrt / primary image 1 (CLI as admin):
   - > boot system image1
   - > reboot

From OpenWrt to stock / boot image 2: (shell as root)
   - # fw_setsys bootpartition 1
   - # reboot

Debrick using serial interface
------------------------------

1. Press "a" "c" "p" during message "Enter correct key to stop autoboot"
2. Load vendor image with "upgrade runtime <TFTP IP>:LGS310xxxxx.imag"
3. switch to primary partition "setsys bootpartition 0"
4. safe config "savesys"

Further documentation
---------------------
See https://openwrt.org/toh/linksys/lgs352c

It has been developed and tested on device with v1 revision.

```

Neustradamus commented:
@plappermaul: Nice PR, hope that it will be merged...

RTL8... is very important...

plappermaul committed:
realtek: add support for Linksys LGS310C

Hardware specification
----------------------

* RTL8380M SoC, 1 MIPS 4KEc core @ 500MHz
* 256MB DRAM
* 32MB NOR Flash
* 8 x 10/100/1000BASE-T ports
* 2 x SFP ports
* Power LED, Fault LED
* Reset button on front panel
* UART (115200 8N1) via populated standard pin header marked JP1

TODO: The SFP ports use a shared SCL GPIO that the driver cannot handle.
The left SFP port (lan9) is defined and fully functional while the laser
on the right SFP port (lan10) is off by default.

UART pinout
-----------

[o]ooo|JP1
 | ||`------ GND
 | |`------- RX
 | `-------- TX
 `---------- Vcc (3V3)

Installation using OEM webinterface
-----------------------------------

1. Make sure you are running OEM firmware in secondary slot
2. Install squashfs-factory.imag to primary slot by upload via http

Installation using serial interface
-----------------------------------

1. Press "a" "c" "p" during message "Enter correct key to stop autoboot"
2. Load image with "upgrade runtime <TFTP IP>:squashfs-sysupgrade.bin" command
3. Switch to primary slot with "setsys bootpartition 0"
4. Store config with "savesys"
5. Boot the image with `boota` command

Dual-boot with stock firmware using writable u-boot-env
-------------------------------------------------------

From stock to OpenWrt / primary image 1 (CLI as admin):
   - > boot system image1
   - > reboot

From OpenWrt to stock / boot image 2: (shell as root)
   - # fw_setsys bootpartition 1
   - # reboot

Debrick using serial interface
------------------------------

1. Press "a" "c" "p" during message "Enter correct key to stop autoboot"
2. Load vendor image with "upgrade runtime <TFTP IP>:LGS310xxxxx.imag"
3. switch to primary partition "setsys bootpartition 0"
4. safe config "savesys"

Further documentation
---------------------
See https://openwrt.org/toh/linksys/lgs352c

It has been developed and tested on device with v1 revision.

Signed-off-by: Markus Stockhausen <markus.stockhausen@gmx.de>

plappermaul commented:
@svanheule rebased on top of current master. With three out-of-tree devices for development a merge will help me a lot.

plappermaul commented:
@svanheule ping

svanheule commented on scripts/linksys-header.py:
```diff
@@ -0,0 +1,89 @@
+    if model == "BKS-RTL83xx":
+        head[0:4] = MAGIC83.to_bytes(4, 'big')
+    else:
+        head[0:4] = MAGIC93.to_bytes(4, 'big')
```
If more models are expected to be added, it would be more flexible to also provide this header magic via the device recipe in the Makefile

plappermaul replied:
Will leave as is. For now only these two types where identified

plappermaul commented on target/linux/realtek/image/rtl838x.mk:
```diff
@@ -153,6 +153,25 @@ define Device/iodata_bsh-g24mb
+  IMAGE_SIZE := 13504k
+  DEVICE_VENDOR := Linksys
+  DEVICE_MODEL := LGS310C
+  LINKSYS_MODEL_ID := BKS-RTL83xx
```
fixed

plappermaul commented on scripts/linksys-image.sh:
```diff
@@ -0,0 +1,71 @@
+gen_imageupgrade() {
+	echo '#!/bin/sh'
+	echo 'flash_bank=65536'
+	echo 'filesize=`ls -la ./series_vmlinux.bix | awk '"'"'{print$5}'"'"'`'
```
works, fixed

plappermaul commented on scripts/linksys-image.sh:
```diff
@@ -0,0 +1,71 @@
+	echo 'flash_bank=65536'
+	echo 'filesize=`ls -la ./series_vmlinux.bix | awk '"'"'{print$5}'"'"'`'
+	echo 'abc=`expr ${filesize} / ${flash_bank} + 1`'
+	echo 'num_bank=${abc}'
```
fixed

plappermaul commented on scripts/linksys-header.py:
```diff
@@ -0,0 +1,89 @@
+VERSION2 = 1
+VERSION3 = 2
+VERSION4 = 2
+COMPANY = "belkin"
```
renamed

plappermaul commented on scripts/linksys-header.py:
```diff
@@ -0,0 +1,89 @@
+
+    return head
+
+parser = argparse.ArgumentParser(description='Generate Linksys LGS3xxC/LGS3xxMPC header.')
```
No idea, leaving as is.

plappermaul commented on scripts/linksys-image.sh:
```diff
@@ -0,0 +1,71 @@
+	echo '#!/bin/sh'
+	echo 'flash_bank=65536'
+	echo 'filesize=`ls -la ./series_vmlinux.bix | awk '"'"'{print$5}'"'"'`'
+	echo 'abc=`expr ${filesize} / ${flash_bank} + 1`'
```
fixed

svanheule commented on scripts/linksys-image.sh:
```diff
@@ -0,0 +1,69 @@
+
+# A file with a list of supported devices. Might contain multiple lines.
+gen_supportlist() {
+	echo 'LGS310C MagicNumber 0x60402060 8380_2fib_1g 3955,1000,201,3,10,1'
```
As you've already written on the wiki, this entry is different for every device in this series. How do you see this scale to other devices?

If you pass $DEVICE_MODEL to the script, in addition to the magic number and extra info (internal name seems to correlate with the upper 6 digits of the magic number), you can put all the required info in the Device definition. Or just pass everything after "MagicNumber" as a single string.

plappermaul replied:
Fixed

svanheule replied:
Reducing the `imageupgrade` script to the minimally required info is a good update, but by now using a lookup table based on `DEVICE_MODEL` in linksys-image.sh and belkin-header.py you now have three locations that need to be updated to add a device. I wouldn't call that part an improvement.

Please update/revert the scripts so that device-unique properties (`BELKIN_MODEL` and e.g. `LINKSYS_MAGIC`) are defined in the makefile. That way there is only one location that needs to be updated to add a new device.

plappermaul replied:
Working alone to get the Target to 6.6 and starting this discussion feels like 2 years ago. I think it is time to stop again.

plappermaul replied:
fixed

svanheule commented on target/linux/realtek/dts/rtl8380_linksys_lgs310c.dts:
```diff
@@ -0,0 +1,212 @@
+				reg = <0x000a0000 0x500000>;
+			};
+			partition@5a0000 {
+				label = "firmare";
```
checkpatch.pl informed me this has a missing 'w'. I fixed it in the merged commit, but I don't think mtdsplit was actually functioning...

plappermaul replied:
thanks

svanheule commented:
Merged into main as of commit d03f3dcf3b2801c0ee8d41506a1da64d3077f19b
